### PR TITLE
feat: rebuild catalog (#3225)

### LINKED ISSUE #3225: Rebuild Catalog
We wish to be able to rebuild the catalog of a database in the event the catalog is lost or corrupted. This functionality is currently implemented `parquet_catalog::rebuild::rebuild_catalog` but is not user-accessible. It should be possible to request that IOx attempts to rebuild the catalog for a Database, if it has failed to load i.e. it is in `DatabaseState::CatalogLoadError`. 

This would be similar to the similar to the existing `WipePreservedCatalog` but would instead run the rebuild logic.

This will also enable building a catalog from a set of parquet files produced by IOx.

* Create new database
* Stop IOx so it doesn't delete "orphaned parquet files"
* Delete transactions directory so catalog load on startup fails
* Add parquet files to `data` directory
* Startup IOx and run a catalog rebuild e.g. `influxdb_iox database recover rebuild`

Split out from #3209

### PR DESCRIPTION
**Pull request by tustvold**
Closes #3225

Adds the ability to rebuild a catalog.

I tested this locally by running a server with

```
cargo run -- run -vvv database --object-store file --data-dir ~/.iox/data/ --server-id 1000 --wipe-catalog-on-error no
```

I then wrote some data into it, and persisted it

```
cargo run database partition persist DB_NAME TABLE_NAME PARTITION_KEY --force
```

I then stopped the server running and nvaigated to the database's directory within `~/.iox/data`. I then ran

```
rm -rf transactions
mkdir -p transactions/00000000000000000000/
echo "INVALID" > transactions/00000000000000000000/78451672-db8c-4f7d-bcca-a4e1d8b5a74d.txn
```

I then started the server up again with the same command and it failed to startup (as expected)

```
2021-11-29T22:21:42.223538Z ERROR server::database: database in error state - operator intervention required db_name=my_database e=error loading catalog: Cannot load preserved catalog: Error during protobuf IO: Error during protobuf deserialization: failed to decode Protobuf message: Transaction.previous_uuid: invalid wire type: SixtyFourBit (expected LengthDelimited) state=CatalogLoadError
```

From a separate terminal I then ran

```
cargo run -- database recover rebuild my_database
```

Then

```
cargo run -- operation wait 0
```

And once that finished I was able to query the database, and all the chunks had been discovered :tada:

### CODE REVIEW

File: influxdb_iox/src/commands/database/recover.rs [hunk 1]
```diff
@@ -72,7 +87,7 @@ pub async fn command(connection: Connection, config: Config) -> Result<()> {
             }
 
             let operation = client
-                .wipe_persisted_catalog(db_name)
+                .wipe_preserved_catalog(db_name)
```
Comment (tustvold) [inline influxdb_iox/src/commands/database/recover.rs:90]: Some parts of the system referred to it as the "persisted catalog" and others the "preserved catalog". I went with the latter as this reflects the structures in parquet_catalog and the gRPC API.

File: influxdb_iox_client/src/client/management.rs [hunk 2]
```diff
@@ -298,7 +298,31 @@ pub enum GetServerStatusError {
 
 /// Errors returned by [`Client::wipe_persisted_catalog`]
 #[derive(Debug, Error)]
-pub enum WipePersistedCatalogError {
+pub enum WipePreservedCatalogError {
```
Comment (tustvold) [inline influxdb_iox_client/src/client/management.rs:301]: #3251 will avoid this silliness

File: parquet_catalog/src/rebuild.rs [hunk 3]
```diff
@@ -65,14 +65,10 @@ pub type Result<T, E = Error> = std::result::Result<T, E>;
 ///
 /// - **Metadata Read Failure:** There is a parquet file with metadata that cannot be read. Set
 ///   `ignore_metadata_read_failure` to `true` to ignore these cases.
-pub async fn rebuild_catalog<S>(
+pub async fn rebuild_catalog(
```
Comment (tustvold) [inline parquet_catalog/src/rebuild.rs:68]: FYI @crepererum I couldn't see a strong reason that this needed to take a CatalogState and so I removed this as it made it easier to use. Please let me know if I'm missing something here

Comment (crepererum) [reply parquet_catalog/src/rebuild.rs:68]: The reason is pretty simple: you probably wanna use the rebuilt catalog straight away. Now you're basically loading it twice.

File: influxdb_iox/tests/common/server_fixture.rs [hunk 4]
```diff
@@ -304,6 +305,20 @@ impl ServerFixture {
     pub fn dir(&self) -> &Path {
         self.server.dir.path()
     }
+
+    /// Poison persisted catalog
+    pub fn poison_catalog(&self, database_uuid: Uuid) {
```
Comment (crepererum) [inline influxdb_iox/tests/common/server_fixture.rs:310]: We already have code for that:

https://github.com/influxdata/influxdb_iox/blob/f165faf5bce152bef6676f4af7fef699c5cfd085/parquet_catalog/src/test_helpers.rs#L234

File: server/src/database.rs [hunk 5]
```diff
@@ -532,6 +542,54 @@ impl Database {
...
+                    Arc::clone(shared.application.time_provider()),
+                );
+                parquet_catalog::rebuild::rebuild_catalog(config, false)
+                    .await
+                    .map_err(Box::new)
+                    .context(RebuildPreservedCatalog { db_name })?;
+
+                {
+                    let mut state = shared.state.write();
+                    *state.unfreeze(handle) = DatabaseState::RulesLoaded(current_state);
```
Comment (crepererum) [inline server/src/database.rs:582]: If you would return the catalog state from `rebuild_catalog` (as it was done before) you could advance the state even further.

File: parquet_catalog/src/rebuild.rs (hunk 3 above)
Comment (tustvold) [reply parquet_catalog/src/rebuild.rs:68]: Is that really a problem? I can change back, but it forces the catalog rebuild to be coordinated by `Db` as `Loader` and friends are private implementation details of it? This seems more complicated?

File: influxdb_iox/tests/common/server_fixture.rs (hunk 4 above)
Comment (tustvold) [reply influxdb_iox/tests/common/server_fixture.rs:310]: This server fixture does not have access to a PreservedCatalog... Its in a separate process

File: parquet_catalog/src/rebuild.rs (hunk 3 above)
Comment (crepererum) [reply parquet_catalog/src/rebuild.rs:68]: Let's be pragmatic then :+1:

**Review by crepererum — APPROVED** [review]

**Comment by alamb** [comment]
FWIW I am going to work on testing this feature by documenting my usecase of "how to rebuild a part of a production cluster"

**Pull request merged**